syntax = "proto3";
package news_hub.sidecar;

import "domain_models.proto";

service SidecarApi {
  rpc GetBoards (GetBoardsReq) returns (GetBoardsRes) {}
  rpc GetThreads (GetThreadsReq) returns (GetThreadsRes) {}
  rpc GetOriginalPost (GetOriginalPostReq) returns (GetOriginalPostRes) {}
  rpc GetReplies (GetRepliesReq) returns (GetRepliesRes) {}
  rpc GetComments (GetCommentsReq) returns (GetCommentsRes) {}

   // Extension operations
  rpc ListInstalledExtensions (news_hub.domain.Empty) returns (ListInstalledExtensionsRes);
  rpc GetInstalledExtension (GetInstalledExtensionReq) returns (GetInstalledExtensionRes);
  rpc InstallExtension (InstallExtensionReq) returns (news_hub.domain.Empty);
  rpc UninstallExtension (UninstallExtensionReq) returns (news_hub.domain.Empty);
  rpc GetInstallProgress (GetInstallProgressReq) returns (GetInstallProgressRes);
  rpc ListRemoteExtensions (ListRemoteExtensionsReq) returns (ListRemoteExtensionsRes);

  // Extension Repository operations
  rpc ListExtensionRepos (news_hub.domain.Empty) returns (ListExtensionReposRes);
  rpc AddExtensionRepo (AddExtensionRepoReq) returns (AddExtensionRepoRes);
  rpc RemoveExtensionRepo (RemoveExtensionRepoReq) returns (news_hub.domain.Empty);

  // Health check
  rpc HealthCheck (HealthCheckReq) returns (HealthCheckRes);
  rpc WatchHealth (HealthCheckReq) returns (stream HealthCheckRes);

  // Logs
  rpc WatchLogs (WatchLogsReq) returns (stream news_hub.domain.LogEntry);
  rpc GetLogs (GetLogsReq) returns (GetLogsRes);
  rpc SetLogLevel (SetLogLevelReq) returns (news_hub.domain.Empty);
}

message GetBoardsReq {
  string pkg_name = 3;
  optional news_hub.domain.PaginationReq page = 2;
}

message GetBoardsRes {
  repeated news_hub.domain.Board boards = 1;
  news_hub.domain.PaginationRes page = 2;
}

message GetThreadsReq {
  string pkg_name = 6;
  string board_id = 8;
  optional string sort = 7;
  optional news_hub.domain.PaginationReq page = 3;
  optional string keywords = 5;
}

message GetThreadsRes {
  repeated news_hub.domain.Post threads = 1;
  news_hub.domain.PaginationRes page = 2;
}

message GetOriginalPostReq {
  string pkg_name = 5;
  string board_id = 3;
  string thread_id = 1;
  string post_id = 4;
}

message GetOriginalPostRes {
  news_hub.domain.Post original_post = 1;
}

message GetRepliesReq {
  string pkg_name = 6;
  string board_id = 2;
  string thread_id = 3;
  optional string reply_to_id = 5;
  optional news_hub.domain.PaginationReq page = 4;
}

message GetRepliesRes {
  repeated news_hub.domain.Post replies = 1;
  news_hub.domain.PaginationRes page = 2;
}

message GetCommentsReq {
  string pkg_name = 6;
  string board_id = 2;
  string thread_id = 3;
  string post_id = 4;
  optional news_hub.domain.PaginationReq page = 5;
}

message GetCommentsRes {
  repeated news_hub.domain.Comment comments = 1;
  news_hub.domain.PaginationRes page = 2;
}

message ListInstalledExtensionsRes {
  repeated news_hub.domain.Extension extensions = 1;
}

message GetInstalledExtensionReq {
  string pkg_name = 1;
}

message GetInstalledExtensionRes {
  news_hub.domain.Extension extension = 1;
}

message InstallExtensionReq {
  string pkg_name = 1;
  optional string repo_url = 4;
}

message UninstallExtensionReq {
  string pkg_name = 1;
}

message GetInstallProgressReq {
  string pkg_name = 1;
}

message GetInstallProgressRes {
  int64 progress = 2;
}

message ListRemoteExtensionsReq {
  optional string keyword = 1;
}

message ListRemoteExtensionsRes {
  repeated news_hub.domain.RemoteExtension extensions = 1;
}

message ListExtensionReposRes {
  repeated news_hub.domain.ExtensionRepo repos = 1;
}

message AddExtensionRepoReq {
  string url = 1;
}

message AddExtensionRepoRes {
  string url = 1;
  int64 added_at = 2;  // milliseconds since epoch
  string display_name = 3;
  string website = 4;
  string signing_key_fingerprint = 5;
  optional string icon = 6;
}

message RemoveExtensionRepoReq {
  string url = 1;
}

message HealthCheckReq {
  string service = 1;
}

message HealthCheckRes {
  enum ServingStatus {
    UNKNOWN = 0;
    SERVING = 1;
    NOT_SERVING = 2;
    SERVICE_UNKNOWN = 3;
  }
  ServingStatus status = 1;
  optional string message = 2;
}

message WatchLogsReq {
  news_hub.domain.LogLevel min_level = 1;
  optional string logger_filter = 2;
}

message GetLogsReq {
  int64 start_time = 1;
  int64 end_time = 2;
  news_hub.domain.LogLevel min_level = 3;
  optional string logger_filter = 4;
  optional int32 limit = 5;
}

message GetLogsRes {
  repeated news_hub.domain.LogEntry entries = 1;
}

message SetLogLevelReq {
  news_hub.domain.LogLevel level = 1;
}
